File: examples/lang/python/src/guguguexamples/codec/json.py
from __future__ import annotations
from dataclasses import dataclass
from typing import List, Callable, Tuple, Optional, TypeVar, Union
from datetime import datetime

from gugugu.lang.python.runtime.codec import (
    EncoderImpl,
    DecoderImpl,
)

JsonRepr = Union[None, int, float, str, list, dict]
A = TypeVar("A")


@dataclass()
class JsonCursor:
    top: JsonRepr
    stack: List[JsonRepr]

    @classmethod
    def from_json(cls, v: JsonRepr) -> JsonCursor:
        return cls(top=v, stack=list())

    def assert_empty_stack(self) -> None:
        if len(self.stack) != 0:
            raise CodecException("unexpected finalize")

    def push(self, new_top: JsonRepr) -> None:
        self.stack.append(self.top)
        self.top = new_top

    def pop(self) -> JsonRepr:
        if len(self.stack) == 0:
            raise CodecException("stack empty")
        old_top = self.top
        self.top = self.stack.pop()
        return old_top


class CodecException(Exception):
    pass


class JsonCodecImpl(EncoderImpl[JsonCursor, JsonRepr],
                    DecoderImpl[JsonCursor, JsonRepr]):

    UNIT = object()

    def encode_with_state(self,
                          k: Callable[[JsonCursor], JsonCursor]) -> JsonRepr:
        s = JsonCursor.from_json(None)
        s = k(s)
        s.assert_empty_stack()
        return s.top

    def decode_with_state(self, r: JsonRepr,
                          k: Callable[[JsonCursor], Tuple[JsonCursor, A]],
                          ) -> A:
        s = JsonCursor.from_json(r)
        s, a = k(s)
        s.assert_empty_stack()
        return a

    def encode_record(self, s: JsonCursor, n_fields: int,
                      k: Callable[[JsonCursor], JsonCursor]) -> JsonCursor:
        s.top = dict()
        return k(s)

    def encode_record_field(self, s: JsonCursor, i: int, name: str,
                            k: Callable[[JsonCursor], JsonCursor],
                            ) -> JsonCursor:
        s.push(None)
        s = k(s)
        value = s.pop()
        if not isinstance(s.top, dict):
            raise CodecException(f"cannot write Record field: {name}")
        s.top[name] = value
        return s

    def decode_record(self, s: JsonCursor, n_fields: int,
                      k: Callable[[JsonCursor], Tuple[JsonCursor, A]],
                      ) -> Tuple[JsonCursor, A]:
        if not isinstance(s.top, dict) or len(s.top) < n_fields:
            raise CodecException("cannot read Record")
        return k(s)

    def decode_record_field(self, s: JsonCursor, i: int, name: str,
                            k: Callable[[JsonCursor], Tuple[JsonCursor, A]],
                            ) -> Tuple[JsonCursor, A]:
        top = s.top
        if not isinstance(top, dict):
            raise CodecException("cannot read Record")
        if name not in top:
            raise CodecException(f"cannot read Record field: {name}")
        s.push(top[name])
        s, a = k(s)
        s.pop()
        return s, a

    def encode_enum(self, s: JsonCursor, a: A, as_index: Callable[[A], int],
                    as_name: Callable[[A], str]) -> JsonCursor:
        return self.encode_string(s, as_name(a))

    def decode_enum(self, s: JsonCursor,
                    by_index: Callable[[int], Optional[A]],
                    by_name: Callable[[str], Optional[A]],
                    ) -> Tuple[JsonCursor, A]:
        s, name = self.decode_string(s)
        a = by_name(name)
        if a is None:
            raise CodecException("cannot read Enum")
        return s, a

    def encode_maybe(self, s: JsonCursor, v: Optional[A],
                     k: Callable[[JsonCursor, A], JsonCursor],
                     ) -> JsonCursor:
        if v is None:
            s.top = None
        else:
            s = k(s, v)
        return s

    def decode_maybe(self, s: JsonCursor,
                     k: Callable[[JsonCursor], Tuple[JsonCursor, A]],
                     ) -> Tuple[JsonCursor, Optional[A]]:
        if s.top is None:
            return s, None
        return k(s)

    def encode_list(self, s: JsonCursor, v: List[A],
                    k: Callable[[JsonCursor, A], JsonCursor],
                    ) -> JsonCursor:
        s.top = list()
        for a in v:
            s.push(None)
            s = k(s, a)
            value = s.pop()
            if not isinstance(s.top, list):
                raise CodecException("bad state: encode_list")
            s.top.append(value)
        return s

    def decode_list(self, s: JsonCursor,
                    k: Callable[[JsonCursor], Tuple[JsonCursor, A]],
                    ) -> Tuple[JsonCursor, List[A]]:
        if not isinstance(s.top, list):
            raise CodecException("cannot decode List")
        rv = list()
        for value in s.top:
            s.push(value)
            s, a = k(s)
            s.pop()
            rv.append(a)
        return s, rv

    def encode_unit(self, s: JsonCursor, v: object) -> JsonCursor:
        s.top = None
        return s

    def decode_unit(self, s: JsonCursor) -> Tuple[JsonCursor, object]:
        if s.top is not None:
            raise CodecException("cannot read Unit")
        return s, self.UNIT

    def encode_bool(self, s: JsonCursor, v: bool) -> JsonCursor:
        s.top = v
        return s

    def decode_bool(self, s: JsonCursor) -> Tuple[JsonCursor, bool]:
        if not isinstance(s.top, bool):
            raise CodecException("cannot read Bool")
        return s, s.top

    def encode_int32(self, s: JsonCursor, v: int) -> JsonCursor:
        s.top = v
        return s

    def decode_int32(self, s: JsonCursor) -> Tuple[JsonCursor, int]:
        if not isinstance(s.top, int):
            raise CodecException("cannot read Int32")
        return s, s.top

    def encode_double(self, s: JsonCursor, v: float) -> JsonCursor:
        s.top = v
        return s

    def decode_double(self, s: JsonCursor) -> Tuple[JsonCursor, float]:
        if not isinstance(s.top, float):
            raise CodecException("cannot read Double")
        return s, s.top

    def encode_string(self, s: JsonCursor, v: str) -> JsonCursor:
        s.top = v
        return s

    def decode_string(self, s: JsonCursor) -> Tuple[JsonCursor, str]:
        if not isinstance(s.top, str):
            raise CodecException("cannot read String")
        return s, s.top

    FORMATTER = "%Y-%m-%dT%H:%M:%S"

    def encode_date_time(self, s: JsonCursor, v: datetime) -> JsonCursor:
        str_ = v.strftime(self.FORMATTER)
        return self.encode_string(s, str_)

    def decode_date_time(self, s: JsonCursor) -> Tuple[JsonCursor, datetime]:
        s, str_ = self.decode_string(s)
        return s, datetime.strptime(str_, self.FORMATTER)

File: examples/lang/haskell/src/GuguguExamples/Codec/Json.hs
{-# OPTIONS_GHC -Wall -Werror #-}
{-# LANGUAGE FlexibleInstances     #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings     #-}
{-# LANGUAGE RecordWildCards       #-}
module GuguguExamples.Codec.Json
  ( JsonCodecImpl(..)
  ) where

import           Control.Monad.Fail
import           Control.Monad.State               hiding (fail)
import           Data.Aeson                        (Value (..))
import qualified Data.HashMap.Strict               as HM
import           Data.List.NonEmpty                (NonEmpty (..), (<|))
import qualified Data.List.NonEmpty                as NonEmpty
import           Data.Scientific
import qualified Data.Text                         as T
import           Data.Time
import           Data.Traversable
import           Prelude                           hiding (fail)
import           Text.Printf

import           Gugugu.Lang.Haskell.Runtime.Codec


data JsonCodecImpl = JsonCodecImpl

type Cursor = NonEmpty Value

data P a
  = Success a
  | Failure String

dateTimeFormat :: String
dateTimeFormat = "%Y-%m-%dT%H:%M:%S"

putFocus :: Value -> StateT Cursor P ()
putFocus v = modify $ \(_ :| vs) -> v :| vs

getFocus :: StateT Cursor P Value
getFocus = gets NonEmpty.head

withPushed :: Value -> StateT Cursor P a -> StateT Cursor P a
withPushed v k = do
  modify (v <|)
  a <- k
  _ :| vs <- get
  case vs of
    []       -> fail "Bad state"
    v' : vs' -> do
      put $ v' :| vs'
      pure a


instance EncoderImpl JsonCodecImpl Value (Either String) (StateT Cursor P) where
  encodeValue c a = case execStateT (encode c a) (Null :| []) of
    Failure e         -> Left e
    Success (r :| []) -> Right r
    Success _         -> Left "unexpected finalize"

  encodeRecord _ k c a = do
    putFocus (Object HM.empty)
    k c a
  encodeRecordField _ name c a = do
    Object o <- getFocus
    encoded <- withPushed Null $ encode c a *> getFocus
    putFocus $ Object $ HM.insert name encoded o

  encodeEnum _ asName c = encodeString c . asName

  encodeMaybe c a = case a of
    Nothing -> putFocus Null
    Just v  -> encode c v
  encodeList c a = do
    encoded <- for a $ \v -> withPushed Null $ encode c v *> getFocus
    putFocus $ Array encoded

  encodeUnit   _ _ = putFocus Null
  encodeBool   _ a = putFocus $ Bool a
  encodeInt32  _ a = putFocus $ Number $ fromIntegral a
  encodeDouble _ a = putFocus $ Number $ fromFloatDigits a
  encodeString _ a = putFocus $ String a

instance DecoderImpl JsonCodecImpl Value (Either String) (StateT Cursor P) where
  decodeValue c r = case runStateT (decode c) (r :| []) of
    Failure e              -> Left e
    Success (a, (_ :| [])) -> Right a
    Success _              -> Left "unexpected finalize"

  decodeRecord _ k c = do
    Object _ <- getFocus
    k c
  decodeRecordField _ name c = do
    Object o <- getFocus
    case HM.lookup name o of
      Just v  -> withPushed v $ decode c
      Nothing -> fail $ printf "cannot read field: %s" name

  decodeEnum _ byName c = do
    s <- decodeString c
    case byName s of
      Just a  -> pure a
      Nothing -> fail $ printf "cannot read enum: %s" s

  decodeMaybe c = do
    j <- getFocus
    case j of
      Null -> pure Nothing
      _    -> decode c
  decodeList c = do
    Array vs <- getFocus
    for vs $ \v -> withPushed v $ decode c

  decodeUnit   _ = do
    Null <- getFocus
    pure ()
  decodeBool   _ = do
    Bool a <- getFocus
    pure a
  decodeInt32  _ = do
    Number v <- getFocus
    case floatingOrInteger v of
      Right a -> pure a
      Left f  -> fail $ printf "cannot read Int32: %f" (f :: Double)
  decodeDouble _ = do
    Number v <- getFocus
    pure $ toRealFloat v
  decodeString _ = do
    String a <- getFocus
    pure a


instance ForeignEncodersImpl JsonCodecImpl (StateT Cursor P) where
  encodeDateTime c =
    encodeString c . T.pack . formatTime defaultTimeLocale dateTimeFormat

instance ForeignDecodersImpl JsonCodecImpl (StateT Cursor P) where
  decodeDateTime c = do
    s <- decodeString c
    -- parseTimeM in time-1.8.0.2 uses Monad.fail instead of MonadFail.fail
    case parseTimeM False defaultTimeLocale dateTimeFormat (T.unpack s) of
      Just v  -> pure v
      Nothing -> fail "cannot read LocalDateTime"


instance Functor P where
  fmap f p = case p of
    Success a -> Success (f a)
    Failure e -> Failure e

instance Applicative P where
  pure = Success
  pf <*> pa = case pf of
    Success f -> fmap f pa
    Failure e -> Failure e

instance Monad P where
  p >>= k = case p of
    Success a -> k a
    Failure e -> Failure e

instance MonadFail P where
  fail = Failure
